074-additional-volumes-support.md

```diff
@@ -0,0 +1,95 @@
+
+## Motivation
+
+Adding support for additional volumes in the CRD will allow users to:
+
+ - Access log files for enhanced monitoring and debugging.
+ - Use additional configurations or secret management tools that require mounting extra volumes.
+
+Improve flexibility in managing Kafka deployments by allowing users to attach custom storage solutions based on their specific requirements.
+
+## Proposal
+
+### Considerations
+
+- CRD Changes: Update the Kafka CRDs to include an `additionalVolumes` field.
+- Validation: Ensure that the additional volumes are validated correctly and do not conflict with existing volume definitions such as `/var/lib/kafka/` and `/tmp`.
+- Documentation: Update the Strimzi documentation to include examples and guidelines on how to use the new `additionalVolumes` field.
+- Security: Ensure that the additional volumes do not introduce security vulnerabilities by validating the configurations.
+- Testing: Comprehensive testing to cover various scenarios and use cases, ensuring that the feature works as expected.
+
+### Possible Delivery Mechanisms
+
+- CRD Updates: Modify the existing CRDs to include the new field.
+  - This will be done by extending the `PodTemplate.java` class to include the new `additionalVolumes` field as a list.
+- Operator Logic: Update the operator logic to handle the additional volumes during the deployment process.
+  - This should be possible to implement in the `KafkaCluster.java` class.
+
+The configuration could look as follows:
+
+```yaml
+    additionalVolumes:
+    - name: example-configmap
+      path: /path/to/mount/volume
+      #subPath: mykey # Add this to mount only a specific key of the configmap/secret
+      configMap:
+        name: config-map-name
+    - name: temp
+      path: /tmp/logs
+      emptyDir: {}
+    - name: example-csi-volume
+      path: /path/to/mount/volume
+      #subPath: "subpath" # Add this to mount the CSI volume at a specific subpath
+      csi:
+        driver: csi-driver-name
+        readOnly: true
+        volumeAttributes:
+          secretProviderClass: example-secret-provider-class
+    - name: example-projected-volume
+      path: /path/to/mount/volume
+      projected:
+        sources:
+          serviceAccountToken
+```
+
+### Supported volumes
+
+The following types of volumes are proposed as initial support. More types could be added later.
+
+- Secret
+- ConfigMap
+- EmptyDir
+- CSI
+- Projected volume
+
+See initial work in this PR draft:
+<https://github.com/strimzi/strimzi-kafka-operator/pull/10099>
+
+## Affected/not affected projects
+
+<https://github.com/strimzi/strimzi-kafka-operator>
+
+## Compatibility
+
+The proposed changes are designed to be forward-compatible. Future updates to the Strimzi Kafka Operator can build upon this feature without breaking existing configurations.

```

Comment:
That is true. As a workaround, we could either enforce users to always use `/mnt` or similar as a root mounting point, or add a warning about not overriding anything managed by Strimzi.

Personally, I would prefer to have the power to mount anything, anywhere, since that would allow users to configure the instances to very specific use-cases.

Reply:
I'm not sure I would want to limit it only to `/mnt`, But I do not think we want to allow anything. We definitely want to prevent using the paths Strimzi is using => for the data, for tmp, for the binaries.

Reply:
@scholzj do you have something specifically in mind that would be prevented by limiting users to `/mnt`? Because it would be simple to implement and simple to explain. If someone later found it wasn't enough we could always allow more in addition.

Reply:
I don't. I guess you are right. If we limit it to something like `/mnt` now, it is easy to open it more later if needed.

Reply:
We will limit it to `/mnt` for now, as discussed in the community call